components/dada-validate/src/validate/validator.rs

```diff
@@ -520,12 +543,26 @@ impl<'me> Validator<'me> {
                 self.add(validated::ExprData::Seq(validated_exprs), expr)
             }
             syntax::ExprData::Return(with_value) => {
-                if let Some(return_expr) = with_value {
-                    let validated_return_expr = self.validate_expr(*return_expr);
-                    return self.add(validated::ExprData::Return(validated_return_expr), expr);
+                match (self.code.return_type, with_value) {
+                    (Some(return_type), None) => {
+                        dada_ir::error!(self.span(expr), "must return something")
+                            .primary_label("must return something")
+                            .secondary_label(return_type, "because return type is not unit")
+                            .emit(self.db);
```

Comment:
Maybe this wording?

```suggestion
                        dada_ir::error!(self.span(expr), "return requires an expression")
                            .primary_label("cannot just have `return` without an expression afterwards")
                            .secondary_label(return_type, "because the function returns a value")
                            .emit(self.db);
```

Probably too wordy.